src/QsCompiler/CompilerOptimizations/Evaluation.fs

```diff
@@ -13,170 +13,181 @@ open Microsoft.Quantum.QsCompiler.SyntaxTree
+            do! setVars callables (s.Lhs, value)
         | QsValueUpdate s ->
             match s.Lhs with
-            | LocalVarTuple vt -> setVarTuple (isLiteral callables) constants (vt, this.evaluateExpression constants s.Rhs) |> Ok
-            | _ -> "Unknown LHS of value update statement: " + (printExpr s.Lhs.Expression) |> CouldNotEvaluate |> Error
-        // FIXME: implement evaluation of conjugations
-        //| QsConjugation s -> NotImplementedException "missing evaluation for conjugation" |> raise
+            | LocalVarTuple vt ->
+                let! value = this.evaluateExpression s.Rhs
+                do! setVars callables (vt, value)
+            | _ -> yield CouldNotEvaluate ("Unknown LHS of value update statement: " + (printExpr s.Lhs.Expression))
         | QsConditionalStatement s ->
-            let firstEval =
-                s.ConditionalBlocks |>
-                Seq.map (fun (ts, block) -> this.evaluateExpression constants ts |> castToBool id, block) |>
-                Seq.tryFind (fst >> function Ok false -> false | _ -> true)
-            match firstEval, s.Default with
-            | Some (Error s, _), _ -> s |> Error
-            | Some (Ok _, block), _ | None, Value block -> this.evaluateScope constants true block.Body
-            | None, Null -> Ok constants
+            let mutable evalElseCase = true
+            for cond, block in s.ConditionalBlocks do
+                let! value = this.evaluateExpression cond >>= castToBool
+                if value then
+                    do! this.evaluateScope block.Body
+                    evalElseCase <- false
+                    do! Break
+            if evalElseCase then
+                match s.Default with
+                | Value block -> do! this.evaluateScope block.Body
+                | _ -> ()
         | QsForStatement stmt ->
-            result {
-                let iterValues = this.evaluateExpression constants stmt.IterationValues
-                let! iterSeq =
-                    match iterValues.Expression with
-                    | RangeLiteral _ when isLiteral callables iterValues -> rangeLiteralToSeq iterValues.Expression |> Seq.map (IntLiteral >> wrapExpr Int) |> Ok
-                    | ValueArray va -> va :> seq<_> |> Ok
-                    | _ -> "Unknown IterationValue in for loop: " + (printExpr iterValues.Expression) |> CouldNotEvaluate |> Error
-                let constantsRef = ref constants
-                for loopValue in iterSeq do
-                    constantsRef := enterScope !constantsRef
-                    constantsRef := defineVarTuple (isLiteral callables) !constantsRef (fst stmt.LoopItem, loopValue)
-                    do! eval constantsRef true stmt.Body
-                    constantsRef := exitScope !constantsRef
-                return !constantsRef
+            let! iterExpr = this.evaluateExpression stmt.IterationValues
+            let! iterSeq = imperative {
```

Comment:
So, the construction and let! here is for the sole purpose of counting?

Reply:
By let!, do you mean the entire Imperative monad? It is used for counting the number of statements evaluated, but it also tracks the state of the currently defined variables, also allows for interrupts at any time, and also gives Break statements to exit loops early.